src/util/datablock/datablock.c

```diff
@@ -0,0 +1,167 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Apache License, Version 2.0,
+* modified with the Commons Clause restriction.
+*/
+
+#include <string.h>
+#include <assert.h>
+#include <stdio.h>
+#include "datablock.h"
+#include "datablock_iterator.h"
+
+#define MAX(a, b) \
+    ((a > b) ? a : b)
+
+// Computes the number of blocks required to accommodate n items.
+#define ITEM_COUNT_TO_BLOCK_COUNT(n) \
+    MAX(1, n / BLOCK_CAP)
+
+// Computes block index from item index.
+#define ITEM_INDEX_TO_BLOCK_INDEX(idx) \
+    (idx / BLOCK_CAP)
+
+// Computes item position within a block.
+#define ITEM_POSITION_WITHIN_BLOCK(idx) \
+    (idx % BLOCK_CAP)
+
+// Get currently active block.
+#define ACTIVE_BLOCK(dataBlock) \
+    dataBlock->blocks[ITEM_INDEX_TO_BLOCK_INDEX(dataBlock->itemCount)]
+
+// Retrieves block in which item with index resides.
+#define GET_ITEM_BLOCK(dataBlock, idx) \
+    dataBlock->blocks[ITEM_INDEX_TO_BLOCK_INDEX(idx)]
+
+//------------------------------------------------------------------------------
+// Block
+//------------------------------------------------------------------------------
+
+Block *_Block_New(size_t itemSize) {
+    assert(itemSize > 0);
+    Block *block = malloc(sizeof(Block) + BLOCK_CAP * itemSize);
```

Comment:
I researched flexible array members and pragmas as we discussed, and I think the allocation you're doing here is perfectly appropriate. Particularly, the `sizeof` behavior we're seeing (16 byte `Block`) is defined by the standard, not an oddity:

> the size of the structure is as if the flexible array member were omitted except that it may have more trailing padding than the omission would imply.
https://wiki.sei.cmu.edu/confluence/display/c/DCL38-C.+Use+the+correct+syntax+when+declaring+a+flexible+array+member

In short, I think this is fine and that pragmas are not called for!